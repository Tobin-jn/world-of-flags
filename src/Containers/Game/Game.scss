$btn-white: #e2e2e2;

.flag-main {
  // background: gainsboro;
  height: 210px;
  margin: 4.5rem auto;
  width: 350px;
}

.flag-image {
  width: 100%;
}

.option-button {
  background: $btn-white;
}