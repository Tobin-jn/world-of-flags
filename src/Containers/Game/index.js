import React, { Component } from 'react';
import { Route, NavLink } from 'react-router-dom';
import './Game.scss';
import Hint from '../../Components/Hint';
import Results from '../../Components/Results';
import { connect } from 'react-redux';
// import flag from '../../images/flags/mexico.png';


export class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      correct: false,
      incorrect: false,
      showHint: false,
      hint: '',
      hintsUsed: 0,
      hintsExhausted: false,
      pointsPossible: 3,
      totalPoints: 0
    }
  }

  checkAnswer = (e) => {
    const { currentCountry } = this.props;
    console.log(e.target.innerText)
    console.log(currentCountry.name)
    const { innerText } = e.target
    if (innerText === currentCountry.name) {
      this.setState({
        correct: true,
        incorrect: false
      })
    } else {
      this.setState({
        incorrect: true,
        correct: false
      })
    }
  }

  giveHint = () => {
    let { hintsUsed, pointsPossible } = this.state;
    const { outline, questions } = this.props.currentCountry;

    this.setState({
      showHint: true
    });
    
    if (hintsUsed === 0) {
      this.setState({
        hint: 'fact'
      });
    } 
    
    if (hintsUsed === 1) {
      this.setState({
        hint: 'outline'
      });
    }

    if (hintsUsed >= 2) {
      this.setState({
        hint: 'out of hints',
        hintsExhausted: true,
      })
    }

    this.setState({
      hintsUsed: hintsUsed + 1,
      pointsPossible: pointsPossible - 1,
    });
  }

  showButtons = () => {
    const { multipleChoice } = this.props.currentCountry

    if(multipleChoice !== undefined) {
      return multipleChoice.map(choice => {
        return (<div 
                  className='option-button button' 
                  key={choice}
                  name={choice}
                  onClick={this.checkAnswer}
                >
                  {choice}
                </div>)  
      })
    }
  }

  closeResults = () => {
    this.setState({
      incorrect: false,
      correct: false,
      hintsExhausted: false,
      hintsUsed: 0,
      pointsPossible: 3,
    });
  }

  addPoints = () => {
    if(this.state.correct === true){
      const points = this.state.pointsPossible
      const totalPoints = this.state.totalPoints + points
      this.setState({ totalPoints })
    }
    else {
      const points = this.state.totalPoints
      this.setState({totalPoints: points})
    }
  }

  hideHint = () => {
    this.setState({
      showHint: false
    });
  }

 getCountryFlagPath = () => {
    const { flag } = this.props.currentCountry;
    const flagUrl = `https://flagz4u.herokuapp.com${flag}`
    return flagUrl
  }

  render() {
    const choiceButtons = this.showButtons();
    const flagImage = this.getCountryFlagPath();

    const { correct, incorrect, pointsPossible, showHint, hint, totalPoints, hintsExhausted } = this.state;
    const { getCountry } = this.props;   
    const { name, facts, country_outline } = this.props.currentCountry;

    return (
      <div className='Game'>
        <div className='top-container'>
          <div className='account-area'>
            <div className='back-button'>
            </div>
            <h5 className='account-text'>Alex <strong>{totalPoints}</strong></h5>
          </div>
        </div>
        <div className='flag-main'>
          <img src={flagImage} className='flag-image'/>
        </div>
        <div className='hint-button' onClick={this.giveHint}>
          Hints: 2
        </div>
        
        { choiceButtons }

        { correct &&
          <Results
            status='Correct'
            closeResults={this.closeResults}
            correctCountry={name}
            points={pointsPossible}
            addPoints={this.addPoints}
            totalPoints={totalPoints}
            hintsExhausted={hintsExhausted}
            getCountry={getCountry}
          />
        }
        { incorrect &&
          <Results
            status='Wrong'
            closeResults={this.closeResults}
            correctCountry={name}
            points={pointsPossible}
            addPoints={this.addPoints}
            totalPoints={this.totalPoints}
            hintsExhausted={hintsExhausted}
            getCountry={getCountry}
          />
        }
        {showHint && 
          <Hint 
            hideHint={this.hideHint} 
            hint={hint}
            outline={country_outline}
            fact={facts.country_fact}
          />
        }
        {/* <Hint /> */}
      </div>
    );
  }
}

export const mapStateToProps = ({ user, currentCountry, usedCountries  }) => ({
  user, 
  currentCountry, 
  usedCountries,   
});

export const mapDispatchToProps = (dispatch) => ({

})

export default connect(mapStateToProps, mapDispatchToProps)(Game);
