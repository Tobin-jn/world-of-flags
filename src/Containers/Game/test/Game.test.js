import React from 'react';
import { shallow, mount } from 'enzyme';
import Game from '../index';

describe('Game', () => {
  let wrapper;
  let mockCountries;
  let mockCorrectCountry;

  beforeEach(() => {
    mockCountries = [
      {
        name: 'Mexico',
        flag: './images/flags/Mexico.png',
        outline: './images/outlines/Mexico.png',
        questions:
          ['Tabasco Hoy',
            'El Financiero & The Oaxaca Times',
            'Oaxaca,Durango,Nayarit']
      },
      {
        name: 'Hungary',
        flag: './images/flags/Hungary.png',
        outline: './images/outlines/Hungary.png',
        questions:
          ['Magyar Nemzet',
            'Its first king is known in English as St. Stephen & locally as Szent Istvan',
            'Nograd,Pest,Somogy']
      },
      {
        name: 'Ireland',
        flag: './images/flags/Ireland.png',
        outline: './images/outlines/Ireland.png',
        questions:
          ['The Connaught Telegraph & The Galway Advertiser',
            'In this country\'s parliament, the Dail has 166 seats, the Seanad, 60']
      },
      {
        name: 'Sweden',
        flag: './images/flags/Sweden.png',
        outline: './images/outlines/Sweden.png',
        questions:
          ['Svenska Dagbladet & Sundsvalls Tidning',
            'Ostergotland,Vasterbotten,Uppsala']
      }
    ];

    mockCorrectCountry = {
      name: 'Sweden',
      flag: './images/flags/Sweden.png',
      outline: './images/outlines/Sweden.png',
      questions:
        ['Svenska Dagbladet & Sundsvalls Tidning',
          'Ostergotland,Vasterbotten,Uppsala']
    };

    wrapper = shallow(<Game 
                        compilePoints={jest.fn()} 
                        correctChoice={mockCorrectCountry} 
                        choices={mockCountries} 
                        totalPoints={10}
                      />);
  });

  it('should render like the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('checkAnswer', () => {
    it.skip('should set state if the answer is correct', () => {
      wrapper.instance().checkAnswer('Sweden');
      expect(wrapper.state().correct).toEqual(true);
    });

    it.skip('should set state if the answer is incorrect', () => {
      wrapper.instance().checkAnswer('Hungary');
      expect(wrapper.state().incorrect).toEqual(true);
    });
  });

  describe('giveHint', () => {
    it('should set state if user requests a hint', () => {
      wrapper.instance().giveHint();
      expect(wrapper.state().hint).toEqual(mockCorrectCountry.questions[0]);
    });

    it('should set state if user requests a second hint', () => {
      wrapper.instance().giveHint();
      wrapper.instance().giveHint();
      expect(wrapper.state().hint).toEqual(mockCorrectCountry.outline);
    });

    it('should exhaust all hints after 2 hints are given', () => {
      wrapper.instance().giveHint();
      wrapper.instance().giveHint();
      wrapper.instance().giveHint();
      expect(wrapper.state().hintsExhausted).toEqual(true);
    });
  });

  describe('mapStateToProps', () => {
    it('should create the correct props object', () => {
    });
  });

  describe('mapDispatchToProps', () => {
    it('should map a key of successfulLogin', () => {
    });

    it('should map a key of signOut', () => {
    });

    it('successfulLogin should call dispatch', () => {
    });

    it('signOut should call dispatch', () => {
    });
  });
});