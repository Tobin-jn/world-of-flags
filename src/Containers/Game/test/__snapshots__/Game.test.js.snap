// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render like the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Game)
    choices={
      Array [
        Object {
          "flag": "./images/flags/Mexico.png",
          "name": "Mexico",
          "outline": "./images/outlines/Mexico.png",
          "questions": Array [
            "Tabasco Hoy",
            "El Financiero & The Oaxaca Times",
            "Oaxaca,Durango,Nayarit",
          ],
        },
        Object {
          "flag": "./images/flags/Hungary.png",
          "name": "Hungary",
          "outline": "./images/outlines/Hungary.png",
          "questions": Array [
            "Magyar Nemzet",
            "Its first king is known in English as St. Stephen & locally as Szent Istvan",
            "Nograd,Pest,Somogy",
          ],
        },
        Object {
          "flag": "./images/flags/Ireland.png",
          "name": "Ireland",
          "outline": "./images/outlines/Ireland.png",
          "questions": Array [
            "The Connaught Telegraph & The Galway Advertiser",
            "In this country's parliament, the Dail has 166 seats, the Seanad, 60",
          ],
        },
        Object {
          "flag": "./images/flags/Sweden.png",
          "name": "Sweden",
          "outline": "./images/outlines/Sweden.png",
          "questions": Array [
            "Svenska Dagbladet & Sundsvalls Tidning",
            "Ostergotland,Vasterbotten,Uppsala",
          ],
        },
      ]
    }
    compilePoints={[MockFunction]}
    correctChoice={
      Object {
        "flag": "./images/flags/Sweden.png",
        "name": "Sweden",
        "outline": "./images/outlines/Sweden.png",
        "questions": Array [
          "Svenska Dagbladet & Sundsvalls Tidning",
          "Ostergotland,Vasterbotten,Uppsala",
        ],
      }
    }
    totalPoints={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
